keps/sig-node/3953-dynamic-node-resize/README.md

```diff
@@ -0,0 +1,776 @@
+This through this both in small and large cases, again with respect to the
+[supported limits].
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+-->
+
+###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?
+
+<!--
+Focus not just on happy cases, but primarily on more pathological cases
+(e.g. probes taking a minute instead of milliseconds, failed pods consuming resources, etc.).
+If any of the resources can be exhausted, how this is mitigated with the existing limits
+(e.g. pods per node) or new limits added by this KEP?
+
+Are there any tests that were run/should be run to understand performance characteristics better
+and validate the declared limits?
+-->
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
```

Comment:
The kubelet talks gRPC to sockets that turn up under `/var/lib/kubelet/device-plugins/`

How about having the kubelet still manage and own these fields within node status, but with a local gRPC interface similar to device plugins.
`/var/lib/kubelet/resource-plugins/` maybe?

This is then an alternative to cAdvisor, and might allow things like:
- a network plugin can advertise the aggregate Pod network throughput to kubelet, which then publishes this somewhere
- a memory manager can advertise the availability of hotpluggable memory as soon as the hotplug is complete (either by announcing the new number, or by prompting kubelet to reevaluate node capacity)
- a CPU manager can advice the kubelet about a hot-unplug event ahead of this happening. The kubelet then notifies the control plane of reduced capacity, triggering cluster autoscaling, descheduling, etc. The reduced capacity might also trigger [node pressure eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/) - all _before_ the CPU actually goes offline.
- following a system event, a node wants to offline 32GiB of memory that is marked unreliable. Similar to the CPU example, an out-of-tree component could notify that intent to the kubelet ahead - perhaps only moments ahead, but still in advance - of configuring the Linux kernel to offline that physical memory region.

One benefit is that hotplug and hot-unplug events might happen during early or late phases of node lifecycle, when network connections to the control plane aren't reliable. A local implementation helps the kubelet use that support to run /docs/concepts/workloads/pods/#static-pods successfully.

Reply:
Both ideas (the main thrust of this KEP, and my alternative in the previous comment) feel compatible. I think server-side apply allows us to set two or more field managers on the status fields for memory and CPU.
For example, `.status.allocatable.cpu` could be co-owned by the kubelet and an external controller, with the kubelet configuration instructing the kubelet to honor the external controller's changes when these are applied.

That field management approach could also work for custom devices. If a device plugin sets capacity for `.status.allocatable."example.com/my-asic"` and `.status.capacity."example.com/my-asic"`  to 8, a controller with access to revise it could change the allocatable figure to 6 (maybe something outside Kubernetes is involved).
Watch out for the work that https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3063-dynamic-resource-allocation/README.md is doing, though.

Reply:
@sftim Thank you sharing information and pointers, Surely we will explore more towards this and keep updated here.

Reply:
@sftim device plugins can dynamically adjusting available capacity using existing ListAndWatch API in DP.

For native resources, we have long standing problem of kubelet assuming information about machine, which might be not real state or might not be the same available to container runtime. In most of the scenarios it is better to trust information provided by runtime (or even runtime handler in some cases) about what is actually available. [KEP-4033](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4033-group-driver-detection-over-cri
) is something that can be used as a basis to pass information (also dynamically updatable) from down to kubelet. Also it can be integrated into data structures similar to [NodeResourceTopology API](https://github.com/kubernetes-sigs/noderesourcetopology-api)

Reply:
We should list some alternatives.

For example, if I hotplug some RAM into a node, the kubelet could respond to that hotplug (no API change needed, mechanism does not need to involve cAdvisor).